feat: `optional` codec and data type (#33)
zarr-developers/zarr-extensions: data-types/optional/README.md

<tr>
<td><code>[42]</code></td>
<td><code>Some(42)</code></td>
</tr>
<tr>
<td rowspan="3"><code>{<br>&nbsp;&nbsp;"name": "optional",<br>&nbsp;&nbsp;"configuration": {<br>&nbsp;&nbsp;&nbsp;&nbsp;"name": "optional",<br>&nbsp;&nbsp;&nbsp;&nbsp;"configuration": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"name": "uint8"<br>&nbsp;&nbsp;&nbsp;&nbsp;}<br>&nbsp;&nbsp;}<br>}</code></td>
<td><code>null</code></td>
<td><code>None</code></td>
</tr>
<tr>
<td><code>[null]</code></td>
<td><code>Some(None)</code></td>
</tr>
<tr>
<td><code>[[42]]</code></td>
<td><code>Some(Some(42))</code></td>
</tr>
</tbody>
</table>

### Python

An `optional` data type with no nesting could be represented using a masked array, such as a NumPy [`numpy.ma.MaskedArray`](https://numpy.org/doc/stable/reference/maskedarray.generic.html).

A `numpy` array using the `StringDType` with an `na_object` that is not `None` could use the `optional` data type with a `string` underlying data type.
However, the `na_object` itself would not be stored in the Zarr metadata of the `optional` data type.
The `na_object` could be set via a runtime option, or alternatively be encoded separately as an attribute, for example.

### Julia

> The Julia `missing` or `NA` values could be used to represent null values in an `optional` data type.

Reviewer: To clarify, the Julia types are [`missing`](https://docs.julialang.org/en/v1/manual/missing/) or [`nothing`](https://docs.julialang.org/en/v1/base/constants/#Core.nothing).

* `missing` is the statistician's null. You can multiply it and perform ternary (Kleene) logic with it.
* `nothing` is the programmer's null or void. It does not participate in computation. It is used similar to `None` in Python.

`NA` is used in the R programming language and is the equivalent of `missing` in Julia.

<td><code>null</code></td>
<td><code>None</code></td>
</tr>
<tr>
<td><code>[42]</code></td>
<td><code>Some(42)</code></td>
</tr>
<tr>
<td rowspan="3"><code>{<br>&nbsp;&nbsp;"name": "optional",<br>&nbsp;&nbsp;"configuration": {<br>&nbsp;&nbsp;&nbsp;&nbsp;"name": "optional",<br>&nbsp;&nbsp;&nbsp;&nbsp;"configuration": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"name": "uint8"<br>&nbsp;&nbsp;&nbsp;&nbsp;}<br>&nbsp;&nbsp;}<br>}</code></td>
<td><code>null</code></td>
<td><code>None</code></td>
</tr>
<tr>
<td><code>[null]</code></td>
<td><code>Some(None)</code></td>
</tr>
<tr>
<td><code>[[42]]</code></td>
<td><code>Some(Some(42))</code></td>
</tr>
</tbody>
</table>

### Python

An `optional` data type with no nesting could be represented using a masked array, such as a NumPy [`numpy.ma.MaskedArray`](https://numpy.org/doc/stable/reference/maskedarray.generic.html).

A `numpy` array using the `StringDType` with an `na_object` that is not `None` could use the `optional` data type with a `string` underlying data type.
However, the `na_object` itself would not be stored in the Zarr metadata of the `optional` data type.

> The `na_object` could be set via a runtime option, or alternatively be encoded separately as an attribute, for example.

Reviewer: ```suggestion
In Python, the representation of `null` values in an `optional` data type may depend on the use of other libraries.
* `None` may be used when using only the Python standard library.
* [`pandas.NA`](https://pandas.pydata.org/docs/reference/api/pandas.NA.html) may be used in conjunction with Pandas.
* [`polars.null`](https://docs.pola.rs/user-guide/expressions/missing-data/) is an appropriate direct mapping for Polars.
* [`pyarrow.null`](https://arrow.apache.org/docs/python/generated/pyarrow.null.html#pyarrow.null) could also be used with pyarrow.
* [`np.nan`](https://numpy.org/doc/2.3/reference/constants.html#numpy.nan) or [`numpy.ma.MaskedArray`](https://numpy.org/doc/stable/reference/maskedarray.generic.html) could be used in NumPy.

A `numpy` array using the `StringDType` with an `na_object` that is not `None` could use the `optional` data type with a `string` underlying data type.
However, the `na_object` itself would not be stored in the Zarr metadata of the `optional` data type.
The `na_object` could be set via a runtime option, or alternatively be encoded separately as an attribute, for example.
```

### Julia

The Julia `missing` or `nothing` values could be used to represent null values in an `optional` data type represented as `Union{T,missing}` or `Union{T,nothing}` where `T` is the underlying data type.

The `null` value type could be set via a runtime option, or alternatively be encoded separately as an attribute, for example.
An `enum`-like codec would be required to distinguish between `missing` or `NA` values within a single array.

> 

Reviewer: ```suggestion
### R

In R, `NA` may be used to represent `null` values in an optional data type.
```

# Optional data type

Defines a data type for optional (nullable) values that can contain either a value of a specified underlying data type or be missing/undefined/null.

> The `optional` data type does not define how null values should be _interpreted_ and represented in-memory.

Reviewer: ```suggestion
The `optional` data type does not define how null values should be _interpreted_, used semantically, or represented in-memory.
```

This is an implementation detail and MUST not impact encoding/decoding.
If an implementation requires a specific in-memory representation for null values (e.g. `missing` or `nothing` in Julia), they are encouraged to do this through a registered attribute.
See the [In-memory representation](#in-memory-representations) section for examples of how null values could be represented in different programming languages.

## Configuration parameters

### `name`
The name of the underlying data type.
This can be any valid Zarr V3 data type name.

### `configuration` (optional)
The configuration object for the underlying data type (if required).
This should match the configuration requirements of the specified underlying data type.

## Permitted fill values

The value of the `fill_value` metadata key MUST be `null` or a single element array containing any valid fill value of the underlying data type.

- A `null` fill value represents the absence of a value (missing element).
- A single element array fill value represents a valid value of the underlying data type.

For nested optional types, this representation is applied recursively.

> ## In-memory representations

Author: It is invalid until it is spec'd. A `bytes` representation would need to be defined for optional data.

I'd rather not include it here and would encourage the split data / mask encoding. However, the `bytes` codec could be supported simply with a tag (present/missing) + payload encoding, always the same size. A more efficient Rust-like [niche optimisation](https://jpfennell.com/posts/enum-type-size/) could be supported for certain inner types, but perhaps that is a bit too fancy?

This data type is designed for the [`optional`](../../codecs/optional/README.md) codec, which separately encodes a validity mask and the data.

While array-to-array codecs MAY support the `optional` data type, implementations SHOULD use the `optional` codec as the sole top-level codec.
This approach is preferred because the codecs contained within the `optional` codec configuration do not need to explicitly handle optional data type semantics.

## Configuration parameters

### `name`
The name of the underlying data type.
This can be any valid Zarr data type name.

### `configuration`
The configuration object for the underlying data type.
This should match the configuration requirements of the specified underlying data type.

## Permitted fill values

The value of the `fill_value` metadata key MUST be `null` or a single element array containing any valid fill value of the underlying data type.

- A `null` fill value represents the absence of a value (missing element).
- A single element array fill value represents a valid value of the underlying data type.

For nested optional types, this representation is applied recursively.

> The table below demonstrates valid `data_type` and `fill_value` combinations with an `optional` and nested `optional` data type, along with their equivalent Rust [`Option`](https://doc.rust-lang.org/std/option/) values.

Author: > It may be necessary to define `null` as its own data type

Why? `null` is just being used as a fill value with specific behaviour for the `optional` data type.

> From the scale-offset codec conversation

Codecs like `scale_offset` could add support for the `optional` data type if there is demand, but I think that is outside the scope of this PR. Inner codecs in `data_codecs` would not interact with the `optional` data type anyway. See below:

> While array-to-array codecs MAY support the `optional` data type, implementations SHOULD use the `optional` codec as the sole top-level codec.
> This approach is preferred because the codecs contained within the `optional` codec configuration do not need to explicitly handle optional data type semantics.

Array-to-array codecs that perform shape manipulation (e.g. [`reshape`](https://github.com/zarr-developers/zarr-extensions/pull/10)) could be an exception here as they support all data types.

> The table below demonstrates valid `data_type` and `fill_value` combinations with an `optional` and nested `optional` data type, along with their equivalent Rust [`Option`](https://doc.rust-lang.org/std/option/) values.

Reviewer: > Why? `null` is just being used as a fill value with specific behaviour for the `optional` data type.

`null` has meaning in other contexts, particularly in other storage contexts. The term arises in SQL, Apache Arrow, and Polars. In the data type fill value description here, we state that `null` indicates the absence of value.

There are two possible ways to interpret an absence of value:
1. The value does not exist.
2. The value is not known.

The first interpretation is typically from the perspective of a software engineer. The value is completely absent. This may be because it was not defined. Trying to access a value that was has not been defined is then thought of as an error that must be handled.

The second interpretation may be from a statistican. The value is missing because we have not measured it or we do not know it. The value may exist, but it is not known to us. Trying to access the value is not an error but rather the lack of knowledge is represented.

Knowing the history of this specification, I would guess that you may mean the first interpretation. In Zarr v2, we also a null fill value meant "no fill value" or the values of an "empty chunk" are undefined. However, in the context of other analogous data libraries, SQL, Apache Arrow, and Polars, `null` means the second interpretation.

Please clarify how `null` should be interpreted here. Leaving the interpretation ambiguous will result in interoperability issues.

> The table below demonstrates valid `data_type` and `fill_value` combinations with an `optional` and nested `optional` data type, along with their equivalent Rust [`Option`](https://doc.rust-lang.org/std/option/) values.

Reviewer: > I don't think zarr usually defines runtime behavior of data, which would mean that discussing the “best” semantics here is probably a distraction. That being said numpy explicitly made it possible to use e.g. `pd.NA` there so you're treating it too harshly.

The issue I'm pointing out is that datatypes do come with some implied runtime behavior. The floating point types rely on IEEE 754 for both their formatting, but this also describes arithmetic on those types. There is implied modular arithmetic on signed and unsigned integers of fixed width.

In this case, the definition of this codec and data type come very close to how [Pola.rs](https://pola.rs/) adresses missing data. For each column, Pola.rs also stores a validity array. In Rust, missing values are also referred to by `Option<T>` when interacting with `Series`.

```rust
use polars::prelude::*;

fn main() -> PolarsResult<()> {
    // Explicitly defining a vector with missing values
    let values: Vec<Option<i32>> = vec![
        Some(10), 
        None,      // This becomes a null in Polars
        Some(30), 
        None
    ];

    let s = Series::new("counts".into(), values);
    
    println!("Series with nulls:\n{}", s);
    Ok(())
}

fn decode_series(s: &Series) {
    // Convert a Series back into a Vec of Options
    // .i32() attempts to view the series as Int32 chunks
    let decoded: Vec<Option<i32>> = s.i32()
        .unwrap()
        .into_iter()
        .collect();

    for val in decoded {
        match val {
            Some(v) => println!("Found value: {}", v),
            None    => println!("Found a missing entry!"),
        }
    }
}
```

However, in Pola.rs,  you can multiply a column with missing data by a scalar.

```rust
use polars::prelude::*;

fn main() -> PolarsResult<()> {
    // 1. Create a DataFrame with Some values and None (nulls)
    let df = df!(
        "sensor_id" => ["A1", "A2", "A3", "A4"],
        "reading" => [Some(10.0), None, Some(25.0), None]
    )?;

    // 2. Multiply the "reading" column by a scalar (e.g., a calibration factor of 1.5)
    // Polars handles the 'None' entries automatically—they stay 'null'.
    let calibrated_df = df.lazy()
        .with_column(
            (col("reading") * lit(1.5)).alias("calibrated_reading")
        )
        .collect()?;

    println!("Resulting DataFrame:\n{}", calibrated_df);

    Ok(())
}
```

This results in the following output:

```
Resulting DataFrame:
shape: (4, 3)
┌───────────┬─────────┬────────────────────┐
│ sensor_id ┆ reading ┆ calibrated_reading │
│ ---       ┆ ---     ┆ ---                │
│ str       ┆ f64     ┆ f64                │
╞═══════════╪═════════╪════════════════════╡
│ A1        ┆ 10.0    ┆ 15.0               │
│ A2        ┆ null    ┆ null               │
│ A3        ┆ 25.0    ┆ 37.5               │
│ A4        ┆ null    ┆ null               │
└───────────┴─────────┴────────────────────┘
```

> The table below demonstrates valid `data_type` and `fill_value` combinations with an `optional` and nested `optional` data type, along with their equivalent Rust [`Option`](https://doc.rust-lang.org/std/option/) values.

Reviewer 2: Apache Arrow defines null as “unknown”, not “missing” https://arrow.apache.org/blog/2022/10/05/arrow-parquet-encoding-part-1/
